github/internals/entities.py

```diff
@@ -0,0 +1,799 @@
+        )
+        created, error = Status.create(self, State.PENDING, description)
+        if not created:
+            return False, "Unable to lock a task '{}': {}".format(
+                self.name, error
+            )
+
+        sleep(RACE_TIMEOUT)
+
+        status = Status.poll(self.pr_number, self.name)
+        if status is None:
+            return False, "Can't get status."
+
+        if status.description != description:
+            return False, "Task '{}' changed. Unable to lock.".format(
+                self.name
+            )
+
+        self.description = description
+
+        return True, None
+
+    def execute(self, statuses, exit_handler=None):
+        """Runs the related task class defined in tasks/tasks.py"""
+        state = State.ERROR
+        description = ""
+        url = ""
+
+        dependencies_results = {}
+        for dep in self.dependencies:
+            status = statuses.get(dep)
+            if status is None:
+                raise RuntimeError(
+                    "Could not check dependencies for task {} PR#{}".format(
+                        self.name, self.pr_number
+                    )
+                )
+            dependencies_results[dep] = JobResult(
+                status.state, status.description, status.target_url
+            )
+
+        try:
+            result = self.job(dependencies_results)
+        except Exception as e:
+            state = State.ERROR
+            description = getattr(
+                e, "description", "{type_}: {msg}".format(
+                    type_=type(e).__name__, msg=str(e)
+                )
+            )
+            url = getattr(e, "url", "")
+            if exit_handler is not None:
+                exit_handler()
+        else:
+            state = result.state
+            description = result.description
+            url = result.url
+        finally:
+            status = Status.poll(self.pr_number, self.name)
+            if status.description != self.description:
+                raise RuntimeError(
+                    "Task {} PR#{} was processed by multiple runners".format(
+                        self.name, self.pr_number
+                    )
+                )
+            created, error = Status.create(self, state, description, url)
+            if not created:
+                raise RuntimeError(error)
+
+
+class ExitHandler(object):
+    done = False
+    aborted = False
+    task = None
+
+    def finish(self, signum, frame):
+        if self.done:
+            return self.abort(signum, frame)
+
+        self.done = True
+
+    def abort(self, signum, frame):
+        if self.aborted:
+            return self.quit()
+
+        self.done = True
+        self.aborted = True
+
+    def quit(self):
+        sys.exit()
+
+    def register_task(self, task):
+        self.task = task
+
+    def unregister_task(self):
+        self.task = None
+
+
+class AbstractJob(Callable):
+    __metaclass__ = ABCMeta
```

Comment:
`python3` disregards `__metaclass__` and still considers this class to be concrete unless it inherits from `ABC`:
`python3`:
```python
>>> class AbstractJob():
...        __metaclass__ = ABCMeta
...        def __init__(self, a, b):
...                self.a = a
...                self.b = b
...        @abstractmethod
...        def x(self):
...            return self.a, self.b
...
>>>
>>> x = AbstractJob()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: __init__() missing 2 required positional arguments: 'a' and 'b'
>>> x = AbstractJob(1,2)
>>> x
<__main__.AbstractJob object at 0x7f12ad2e15c0>
```
whereas `python2`:
```python
>>> class AbstractJob():
...        __metaclass__ = ABCMeta
...        def __init__(self, a, b):
...                self.a = a
...                self.b = b
...        @abstractmethod
...        def x(self):
...            return self.a, self.b
...
>>> x = AbstractJob()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Can't instantiate abstract class AbstractJob with abstract methods x

```